### ISSUE TITLE
Proposal: many_as_possible macro

### ISSUE THREAD

**Issue by zekesonxx** [Geal/nom#558]
So, three wasted hours later, the issue is finally revealed. In the code `flat_map!(take!(main.file_offset), many1!(subheader))`, many1 is not okay with incomplete data at the end. Except, I have all the data, and I need it to ignore random data at the end.

Example:

```rust
named!(testo<Vec<&[u8]>>, many0!(tag!("abcd")));
```

Actual behavior:

```rust
println!("{:?}", testo(&b"abcdabcdef"[..])); // => Done([101, 102], [[97, 98, 99, 100], [97, 98, 99, 100]])
println!("{:?}", testo(&b"abcdabcdab"[..])); // => Incomplete(Size(12))
```

Desired behavior by utilizing a many_as_possible function of some form:

```rust
println!("{:?}", testo(&b"abcdabcdef"[..])); // => Done([101, 102], [[97, 98, 99, 100], [97, 98, 99, 100]])
println!("{:?}", testo(&b"abcdabcdab"[..])); // => Done([99, 100], [[97, 98, 99, 100], [97, 98, 99, 100]])
```

The manually done one I have for the parser I'm actually working on is:

```rust
fn many_as_possible(i: &[u8]) -> IResult<&[u8], Vec<DDSubFileHeader>> {
    let mut output = vec![];
    let mut i = i;
    loop {
        match subheader(i) {
            Error(err) => return Error(err),
            Incomplete(needed) => return Done(i, output),
            Done(newi, file) => {
                i = newi;
                output.push(file);
            }
        }
    }
    Done(i, output)

}
```

TL;DR: `many_as_possible!()`: like `many1!()`, but instead of propagating an Incomplete, return the built Vec up to that point along with the remainder of the input stream.

**Comment by Geal** [comment]
why not use `complete!(tag!("abcd"))`  ?

**Comment by zekesonxx** [comment]
Because the documentation says that `complete!()` converts Incompletes to Errors, and nowhere (outside of reading the code) does it say `many1!()` returns the array on an error instead of propagating the error.
